_posts/2020-03-19-our-rails-upgrade-process.markdown

```diff
@@ -0,0 +1,83 @@
+
+The good news is that here at [Ombu Labs](https://www.ombulabs.com/), we have been upgrading Rails applications for over 10 years now and this gave us a know-how to define a process that has proven to be very effective. And today I want to share Our Rails Upgrade Process with you.
+
+<!--more-->
+
+## Step 1: Finding deprecation warnings
+
+The first thing that we do is to search for deprecation warnings in the application, which are log messages suggesting things that you will need to change when moving for the next version of Rails. That can be because something will either be removed from the source code or the code will work in a different way.
+
+You might find these deprecation warnings by searching at the `production.log` or tracking then down in your Log Management Software, such as [Sentry](https://sentry.io/) or [Splunk](https://www.splunk.com).
+
+For each different deprecation warning that we find, we create one story in our backlog software (at Ombu Labs we use [Pivotal Tracker](pivotaltracker.com) to help us with that). This will make things easier to code review and to test in QA.
+
+## Step 2: Fixing all deprecation warnings
+
+After create all stories for the deprecation warnings, is time to address them one by one. Deprecation warnings are usually pretty straightforward, the messages are clear about what needs to change. They are also backwards compatible which means that you can fix them and merge directly to the current version of Rails.
+
+A good examples is the deprecation warnings caused by the changes in `ActiveRecord::Dirty` in the Rails 5.2 version. We wrote [this article](https://www.fastruby.io/blog/rails/upgrades/active-record-5-1-api-changes.html) while ago showing how to address them.
+
+## Step 3: Add dual boot for the Rails version
+
+To help us switch between the current Rails version and the new one, we usually create a dual boot mechanism. The fastest way is to install the handful gem [next_rails](https://github.com/fastruby/next_rails). Please, [visit this article](https://www.fastruby.io/blog/upgrade-rails/dual-boot/dual-boot-with-rails-6-0-beta.html) where we showed how you can install and setup the gem in your local environment and your CI server.
+
+
+## Step 4: Assessing whether we can upgrade a dependency or not
+
+After shipping all deprecation warnings fixes to the master branch and setup the dual boot, it's time for us to handle with the project dependencies.
+
+Sometimes dependencies are backwards compatible with the current version of Rails. They will have a code like:
+
+```
+if Rails::VERSION >= 5.1
+  # does X
+else
+  # does Y
+end
+```
+
+If that is the case, then you might be able to just upgrade the dependency using `bundle update`.
+
+If the dependency does not have backwards compatible code, then, using the next_rails gem, you can add something like this to your application's Gemfile:
+
+```
+if next?
+  gem "dep", "~> 1.2.3"
+else
+  gem "dep", "~> 0.9"
+end
+```
```

Comment:
Maybe you can briefly explain how `next_rails` works? Maybe you can show the method definition for `next?` and explain that it uses a symlink called `Gemfile.next` in order to keep a separate lockfile: Gemfile.next.lock

Reply:
Also, when a dependency/library has been abandoned, you will have four options to move forward: 

- find an alternative library
- write your own code to address the library’s function and remove library/dependency 
- submit a contribution to the library to add support for the version of Rails that you need
- if your contribution is never merged, start maintaining a fork of the library